### PR TITLE
Refactor polyglot piranha to lower the coupling between the modules

### PR DESCRIPTION
**Pull request by ketkarameya**
This step 0 for the large refactoring that organizes the codebase into cargo workspace. #271 
As part of this cleanup I have : 
1. Moved functions from `Rule` to `SourceCodeUnit` 
2. Replace get/set with macro
3. Remove redundancy in `scope_tests.rs`

(I have annotated the diff)

### CODE REVIEW

File: polyglot/piranha/src/models/unit_tests/scopes_test.rs [hunk 1]
```diff
@@ -20,43 +20,81 @@ use {
   std::{collections::HashMap, path::PathBuf},
 };
 
+struct TestScopesGenerators {
```
Comment (ketkarameya) [inline polyglot/piranha/src/models/unit_tests/scopes_test.rs:23]: I have removed the redundancy from the tests.

Comment (ketkarameya) [reply polyglot/piranha/src/models/unit_tests/scopes_test.rs:23]: Replace this awkward struct with functions :|

File: polyglot/piranha/src/lib.rs [hunk 2]
```diff
@@ -107,245 +90,20 @@ pub fn execute_piranha(
...
-      total_number_of_matches += number_of_matches;
-    }
-    info!("Total files affected/matched {}", &summaries.len());
-    info!("Total number of matches {}", total_number_of_matches);
-    info!("Total number of rewrites {}", total_number_of_rewrites);
-}
-
-impl SourceCodeUnit {
-  /// Will apply the `rule` to all of its occurrences in the source code unit.
-  fn apply_rule(
```
Comment (ketkarameya) [inline polyglot/piranha/src/lib.rs:128]: Moved all functions in this `impl` into the `impl` in `source_code_unit.rs`

File: polyglot/piranha/src/utilities/tree_sitter_utilities.rs [hunk 3]
```diff
@@ -107,25 +105,6 @@ pub(crate) trait PiranhaHelpers {
 }
 
 impl PiranhaHelpers for Node<'_> {
-  /// Checks if the given rule satisfies the constraint of the rule, under the substitutions obtained upon matching `rule.query`
```
Comment (ketkarameya) [inline polyglot/piranha/src/utilities/tree_sitter_utilities.rs:110]: Moved to `rule.rs` as a new trait.

File: polyglot/piranha/src/models/source_code_unit.rs [hunk 4]
```diff
@@ -258,40 +502,95 @@ impl SourceCodeUnit {
...
-  }
-
   pub(crate) fn add_to_substitutions(
     &mut self, new_entries: &HashMap<String, String>, rule_store: &mut RuleStore,
   ) {
     let _ = &self.substitutions.extend(new_entries.clone());
     rule_store.add_global_tags(new_entries);
   }
 
-  pub(crate) fn rewrites(&self) -> &[Edit] {
```
Comment (ketkarameya) [inline polyglot/piranha/src/models/source_code_unit.rs:277]: Replaced these kind of getters with the `macro`

File: polyglot/piranha/src/models/source_code_unit.rs [hunk 5]
```diff
@@ -258,43 +502,127 @@ impl SourceCodeUnit {
...
 
+
+pub(crate) trait SatisfiesConstraint {
+  // / Checks if the given rule satisfies the constraint of the rule, under the substitutions obtained upon matching `rule.query`
+  fn is_satisfied(&self, source_code_unit: &SourceCodeUnit, rule: &Rule, substitutions: &HashMap<String, String>,rule_store: &mut RuleStore,) -> bool ;
+}
+
+/// Checks if the given rule satisfies the constraint of the rule, under the substitutions obtained upon matching `rule.query`
+impl SatisfiesConstraint for Node<'_> {
+fn is_satisfied(
```
Comment (lazaroclapp) [inline polyglot/piranha/src/models/source_code_unit.rs:605]: Indentation is a bit weird here. Shouldn't `is_satisfied` be indented with respect to the impl block containing it? Or is this a quirk about rust formatting I am not aware of?

File: polyglot/piranha/src/models/unit_tests/scopes_test.rs [hunk 6]
```diff
@@ -20,43 +20,68 @@ use {
...
-    )],
-  );
+fn _get_class_scope() -> ScopeGenerator {
+  let scope_query_generator_class: ScopeQueryGenerator = ScopeQueryGeneratorBuilder::default()
+    .matcher("(class_declaration name:(_) @n) @c".to_string())
+    .generator(
+      "(
+    ((class_declaration name:(_) @z) @qc)
+    (#eq? @z \"@n\")
+    )"
```
Comment (lazaroclapp) [inline polyglot/piranha/src/models/unit_tests/scopes_test.rs:30]: Indentation is a bit worse in this version for reading the pattern inline...

File: polyglot/piranha/src/models/unit_tests/scopes_test.rs [hunk 7]
```diff
@@ -20,43 +20,68 @@ use {
...
+                name : (_) @z
+                parameters : (formal_parameters)@tp))
+        (#eq? @z \"@n\")
+        (#eq? @tp \"@fp\")                  
+        )
+       (((constructor_declaration 
+                name: (_) @z
+                parameters : (formal_parameters)@tp))
+        (#eq? @tp \"@fp\")
+        (#eq? @z \"@n\")
```
Comment (lazaroclapp) [inline polyglot/piranha/src/models/unit_tests/scopes_test.rs:67]: Why is the order flipped in this case?

**Review by lazaroclapp — APPROVED** [review]
Refactoring makes sense to me! Very much optional nits about indentation and a question on the DSL test case templates below 🙂

File: polyglot/piranha/src/models/source_code_unit.rs (hunk 5 above)
Comment (ketkarameya) [reply polyglot/piranha/src/models/source_code_unit.rs:605]: I will format this. No I didnt run rust format. :|

File: polyglot/piranha/src/models/unit_tests/scopes_test.rs (hunk 7 above)
Comment (ketkarameya) [reply polyglot/piranha/src/models/unit_tests/scopes_test.rs:67]: no particular reason. I will address this.

**Pull request merged**